traverse/visitor.go

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+Package traverse provides a visitor framework for analyzing Substrait plans
+with multiple read-only operations in a single tree traversal.
+
+## Key Features:
+- Interface-based visitor pattern
+- Single tree walk for multiple analyses
+- Sequential execution
+- Read-only operations - NO MUTATIONS
+- Zero allocations per node
+
+## IMPORTANT: Read-Only Analysis Only
+This interface is designed for read-only analysis, and assumes the tree does not change
+as it is being walked. Any modification of the tree could disrupt the walking of the
+tree causing nodes to be missed, visited out of order, or cause panics.
+
+## Basic Usage:
+
+	// Option 1: Simple usage with Visit (builds context automatically)
+	visitor := NewMyVisitor()
+	traverse.Visit(plan, visitor)
+
+	// Option 2: Explicit context for advanced usage
+	ctx := traverse.NewPlanContext(plan)
+	visitor := NewMyVisitor(ctx)
+	traverse.Walk(plan.Relations, visitor)
+
+	// Option 3: With cycle detection for DAG enforcement
+	visitor := NewMyVisitor()
+	cycleDetector := traverse.NewCycleDetectingVisitor(visitor)
+	traverse.Walk(rel, cycleDetector)
+
+## Creating Custom Visitors:
+
+	// Visitor that only cares about relations
+	type NodeCounter struct {
+		count int
+	}
+
+	func (v *NodeCounter) VisitRel(rel *proto.Rel) {
+		v.count++
+	}
+
+	// Visitor that only cares about expressions (e.g., finding all functions)
```

Comment:
Presumably for finding all utilized functions versus checking the extension list.